### PR TITLE
Add nftables ingress and egress

### PR DESCRIPTION
**Pull request by alphadose**

### CODE REVIEW

**Review by mattkasun — APPROVED** [review]

**Review by Aceix — APPROVED** [review]

File: nmproxy/router/nftables_linux.go [hunk 1]
```diff
@@ -0,0 +1,1396 @@
...
+				UserData: []byte(genRuleKey("-j", "RETURN")),
+			},
+			rule:  []string{"-j", "RETURN"},
+			table: defaultNatTable,
+			chain: netmakerNatChain,
+		},
+	}
+)
+
+// CreateChains - creates default chains and rules
```
Comment (0xdcarns) [inline nmproxy/router/nftables_linux.go:135]: comment form should be nftablesManager.CreateChains -

File: nmproxy/router/nftables_linux.go [hunk 2]
```diff
@@ -0,0 +1,1396 @@
...
+
+	if err := n.conn.Flush(); err != nil {
+		return err
+	}
+	// add jump rules
+	n.addJumpRules()
+	return nil
+}
+
+// CleanRoutingRules cleans existing nftable resources that we created by the agent
```
Comment (0xdcarns) [inline nmproxy/router/nftables_linux.go:231]: see above

File: nmproxy/router/nftables_linux.go [hunk 3]
```diff
@@ -0,0 +1,1396 @@
...
+				if err := n.deleteRule(rule.table, rule.chain, genRuleKey(rule.rule...)); err != nil {
+					logger.Log(0, "Error cleaning up rule: ", err.Error())
+				}
+			}
+		}
+	}
+}
+
+// DeleteRuleTable - deletes all rules from a table
+func (n *nftablesManager) DeleteRuleTable(server, ruleTableName string) {
```
Comment (0xdcarns) [inline nmproxy/router/nftables_linux.go:249]: see above

File: nmproxy/router/nftables_linux.go [hunk 4]
```diff
@@ -0,0 +1,1396 @@
...
+	switch ruleTableName {
+	case ingressTable:
+		delete(n.ingRules, server)
+	case egressTable:
+		delete(n.engressRules, server)
+	}
+}
+
+// InsertEgressRoutingRules - inserts egress routes for the GW peers
+func (n *nftablesManager) InsertEgressRoutingRules(server string, egressInfo models.EgressInfo) error {
```
Comment (0xdcarns) [inline nmproxy/router/nftables_linux.go:262]: see above

File: nmproxy/router/nftables_linux.go [hunk 5]
```diff
@@ -0,0 +1,1396 @@
...
+					})
+			}
+		}
+	}
+	ruleTable[egressInfo.EgressID].rulesMap[egressInfo.EgressID] = egressGwRoutes
+
+	return nil
+}
+
+// AddEgressRoutingRule - inserts an nftable rule for gateway peer
```
Comment (0xdcarns) [inline nmproxy/router/nftables_linux.go:586]: see above

File: nmproxy/router/nftables_linux.go [hunk 6]
```diff
@@ -0,0 +1,1396 @@
...
+					chain:  netmakerFilterChain,
+					rule:   ruleSpec,
+				})
+		}
+	}
+	return nil
+}
+
+// AddIngressRoutingRule - adds a ingress route for a peer
+func (n *nftablesManager) AddIngressRoutingRule(server, extPeerKey, extPeerAddr string, peerInfo models.PeerRouteInfo) error {
```
Comment (0xdcarns) [inline nmproxy/router/nftables_linux.go:708]: see above

File: nmproxy/router/nftables_linux.go [hunk 7]
```diff
@@ -0,0 +1,1396 @@
...
+			nfRule: rule,
+			rule:   ruleSpec,
+			chain:  netmakerFilterChain,
+			table:  defaultIpTable,
+		},
+	}
+	return nil
+}
+
+// InsertIngressRoutingRules inserts an nftables rules for an ext. client to the netmaker chain and if enabled, to the nat chain
```
Comment (0xdcarns) [inline nmproxy/router/nftables_linux.go:805]: see above

File: nmproxy/router/nftables_linux.go [hunk 8]
```diff
@@ -0,0 +1,1396 @@
...
+					rule.table, rule.rule, srcPeerKey, err)
+			}
+		}
+	} else {
+		return errors.New("rules not found for: " + dstPeerKey)
+	}
+	return nil
+}
+
+// FlushAll - removes all the rules added by netmaker and deletes the netmaker chains
```
Comment (0xdcarns) [inline nmproxy/router/nftables_linux.go:1296]: address all public func comments as seen above

**Review by 0xdcarns — CHANGES_REQUESTED** [review]

**Comment by theguy951357** [comment]
will wait on changes

**Review by 0xdcarns — APPROVED** [review]

**Review by abhishek9686 — APPROVED** [review]

**Comment by theguy951357** [comment]
I cannot ping ext client from other server.
Screenshot 2023-02-14 at 1 19 05 PM

**Review by theguy951357 — APPROVED** [review]
can ping ext client in an ingress gateway
can ping network from ext client.
can ping ext client from another ext client
can ping egress gateway
pr checklist passes all points
pass

**Pull request merged**